src/main/java/frc/robot/subsystems/TurretSubsystem.java

```diff
@@ -0,0 +1,246 @@
+
+    simEncoder =
+        new SimDeviceSim("CANEncoder:CANCoder (v6)", Constants.TurretConstants.CANcoder_ID);
+    simTurretPos = simEncoder.getDouble("rawPositionInput");
+
+    zeroTurretEncoder();
+    turretPIDController.setIZone(Constants.TurretConstants.KIZ);
+  }
+
+  public void stopTurret() {
+    turretMotor.stopMotor();
+  }
+
+  public Command stowTurret() {
+    return new InstantCommand(() -> setTurretPos(Math.toRadians(0))).withName("StowTurretCommand");
+  }
+
+  public void setTurretPos(double setpoint) {
+    turretGoalPositionEntry.log(setpoint);
+    turretMotor.set(
+        -turretPIDController.calculate(
+            turretPosition,
+            MathUtil.clamp(
+                setpoint,
+                Math.toRadians(-Constants.TurretConstants.MAX_TURRET_ANGLE_DEGREES),
+                Math.toRadians(Constants.TurretConstants.MAX_TURRET_ANGLE_DEGREES))));
+  }
+
+  public void holdPosition() {
+    setTurretPos(turretPosition);
+  }
+
+  public void zeroTurretEncoder() {
+    m_encoder.setPosition(0.0);
+  }
+
+  public void updateTurretPosition() {
+    turretPosition =
+        Math.toRadians(
+            ((m_encoder.getPosition().getValueAsDouble()) * 360)
+                * Constants.TurretConstants.CONVERSION_FACTOR);
+    SmartDashboard.putNumber("Turret Position", turretPosition);
+    SmartDashboard.putBoolean("Out of Bounds", Math.abs(turretPosition) > 3.14);
+    SmartDashboard.putBoolean(
+        "Soft limit enabled forward", turretMotor.isSoftLimitEnabled(SoftLimitDirection.kForward));
+    SmartDashboard.putBoolean(
+        "Soft limit enabled reverse", turretMotor.isSoftLimitEnabled(SoftLimitDirection.kReverse));
+  }
+
+  public double getTurretPos() {
+    return turretPosition; // returns the absolute encoder position in radians
+  }
+
+  public double getTurretVel() {
+    return turretVelocity;
+  }
+
+  public Command testTurretCommand() {
+    return runEnd(() -> setTurretPos(Math.toRadians(-60)), this::stopTurret);
+  }
+
+  public void AimTurret() {
+    double limelightTX = LimelightHelpers.getTX("limelight");
+    if (limelightTX != 0
+        && LimelightHelpers.getFiducialID("limelight")
+            == ((robotStateManager.getAllianceColor() == AllianceColor.BLUE)
+                ? Constants.TurretConstants.SPEAKER_TAG_ID_BLUE
+                : Constants.TurretConstants.SPEAKER_TAG_ID_RED)) {
+      // X & Rotation
+      setTurretPos(Math.toRadians(limelightTX) + turretPosition);
+
+      // Y & Tilting
+      double limelightTY = LimelightHelpers.getTY("limelight");
+      double distanceToTag = tyToDistanceFromTag(limelightTY);
+      tagDistanceEntry.log(distanceToTag);
+      // TODO: Add vertical tilt and use distance for it
+
+      if (Math.abs(Math.toRadians(limelightTX) + turretPosition)
+          > Math.toRadians(Constants.TurretConstants.MAX_TURRET_ANGLE_DEGREES)) {
+        // TODO: Make turret rotate the drivebase if necessary and driver thinks it's a good idea
```

Comment:
This is probably better done as a composite Command setup somewhere else, not in this subsystem